src/write_log.c

```diff
@@ -56,21 +59,111 @@ static int wl_write_messages (const data_set_t *ds, const value_list_t *vl)
+    size_t bfill = 0;
+
+    if (0 != strcmp (ds->type, vl->type))
+    {
+        ERROR ("write_log plugin: DS type does not match value list type");
+        return -1;
+    }
+
+    format_json_initialize(buffer, &bfill, &bfree);
+    format_json_value_list(buffer, &bfill, &bfree, ds, vl,
+                           /* store rates = */ 0);
+    format_json_finalize(buffer, &bfill, &bfree);
+
+    INFO ("write_log values:\n%s", buffer);
+
+    return (0);
+} /* int wl_write_json */
 
 static int wl_write (const data_set_t *ds, const value_list_t *vl,
-        __attribute__ ((unused)) user_data_t *user_data)
+        user_data_t *user_data)
 {
-    int status;
+    int status = 0;
+    int mode = (int) (size_t) user_data->data;
 
-    status = wl_write_messages (ds, vl);
+    if (mode == WL_FORMAT_GRAPHITE)
+    {
+        status = wl_write_graphite (ds, vl);
+    }
+    else if (mode == WL_FORMAT_JSON)
+    {
+        status = wl_write_json (ds, vl);
+    }
 
     return (status);
 }
 
+static int wl_config (oconfig_item_t *ci) /* {{{ */
+{
+    int mode = 0;
+    for (int i = 0; i < ci->children_num; i++)
+    {
+        oconfig_item_t *child = ci->children + i;
+
+        if (strcasecmp ("Format", child->key) == 0)
+        {
+            char *mode_str = NULL;
+            if ((child->values_num != 1)
+                || (child->values[0].type != OCONFIG_TYPE_STRING))
+            {
+                ERROR ("write_log plugin: Option `%s' requires "
+                    "exactly one string argument.", child->key);
+                return (-EINVAL);
+            }
+            if (mode != 0)
+            {
+                WARNING ("write_log plugin: Redefining option `%s'.",
+                    child->key);
+            }
+            mode_str = child->values[0].value.string;
+            if (strcasecmp ("Graphite", mode_str) == 0)
+                mode = WL_FORMAT_GRAPHITE;
+            else if (strcasecmp ("JSON", mode_str) == 0)
+                mode = WL_FORMAT_JSON;
+            else
+            {
+                ERROR ("write_log plugin: Unknown mode `%s' for option `%s'.",
+                    mode_str, child->key);
+                return (-EINVAL);
+            }
+        }
+        else
+        {
+            ERROR ("write_log plugin: Invalid configuration option: `%s'.",
+                child->key);
+        }
+    }
+    if (mode == 0)
+        mode = WL_FORMAT_GRAPHITE;
+
+    user_data_t ud = {
+        .data = (void *) (size_t) mode,
+        .free_func = NULL
+    };
+
+    plugin_register_write ("write_log", wl_write, &ud);
+
+    return (0);
+} /* }}} int wl_config */
+
 void module_register (void)
 {
-    plugin_register_write ("write_log", wl_write, NULL);
+    plugin_register_complex_config ("write_log", wl_config);
+
+    user_data_t ud = {
+        .data = (void *) (size_t) WL_FORMAT_GRAPHITE,
+        .free_func = NULL
+    };
+
+    plugin_register_write ("write_log", wl_write, &ud);
```

Comment:
@igorpeshansky You're right, I hadn't considered that duplicate registration with the same name would overwrite a previous registration – I was under the impression that the plugin would be called twice when a config was present.

There are a couple of plugins that register write callbacks from within the config callback, but they all *need* there to be a configuration in order to function (and usually allow to configure more than one instance of the plugin). By *moving* the registration to the config handler (and removing it from `module_register()`) we'd *require* a configuration, breaking backwards compatibility.

I'd prefer to call `plugin_register_write()` once from `module_register()` and let the *config* callback set a global variable: no other plugin is calling `plugin_register_*()` with the same name twice *on purpose*, so this behavior is not intuitive. The warning that is logged in that case exists because this is unexpected behavior. Also, passing configuration via a `user_data_t` even though there is only one instance of the plugin is unusual and therefore confusing. Other plugins with only one registered callback simply ignore that user data argument.

Best regards,
—octo